// @cucumber/pretty-formatter/dist/ProgressBarPrinter.js
import { TestStepResultStatus, } from '@cucumber/messages';
import { Query } from '@cucumber/query';
import { composeGlobalHookSummary, composeScenarioSummary, composeSnippets, composeStats, } from './composition/index.js';
import { defaultFormatCode } from './defaultFormatCode.js';
import { formatProblem } from './formatting/formatProblem.js';
import { formatError, formatUndefinedParameterType } from './formatting/index.js';
import { InterferenceInterceptor } from './InterferenceInterceptor.js';
import { findAllSuggestions } from './queries/index.js';
import { CUCUMBER_THEME } from './theme.js';
import { indent, indentNumbered, NON_REPORTABLE_STATUSES, ProblemType } from './utils.js';
var Phase;
(function (Phase) {
    Phase[Phase["PREPARING"] = 0] = "PREPARING";
    Phase[Phase["RUNNING"] = 1] = "RUNNING";
    Phase[Phase["DONE"] = 2] = "DONE";
})(Phase || (Phase = {}));
const DEFAULT_OPTIONS = {
    formatCode: defaultFormatCode,
    includeAttachments: true,
    interference: { mode: 'passthrough' },
    summarise: false,
    theme: CUCUMBER_THEME,
};
const MAX_BAR_WIDTH = 50;
const MIN_LEGEND_WIDTH = 30;
/**
 * Prints test progress as an updating progress bar, with problems prepended above
 *
 * @remarks
 * Requires a TTY terminal.
 */
export class ProgressBarPrinter {
    stream;
    interceptor;
    options;
    query = new Query();
    printedProblems = [];
    pendingProblems = [];
    phase = Phase.PREPARING;
    finishedScenarios = 0;
    finishedSteps = 0;
    runningScenarios = 0;
    totalScenarios = 0;
    totalSteps = 0;
    /**
     * Creates a new ProgressBarPrinter instance
     *
     * @param params - Initialisation object
     * @param params.stream - The stream being written to, used for feature detection
     * @param params.options - Options for the output
     */
    constructor({ stream = process.stdout, options = {}, } = {}) {
        this.stream = stream;
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options,
        };
        this.interceptor = new InterferenceInterceptor(this.options.interference);
    }
    /**
     * Processes a Cucumber message envelope and prints if appropriate
     *
     * @param message - The Cucumber message envelope to process
     */
    update(message) {
        this.query.update(message);
        if (this.phase === Phase.DONE) {
            return;
        }
        if (message.undefinedParameterType) {
            this.handleUndefinedParameterType(message.undefinedParameterType);
        }
        if (message.testRunStarted) {
            this.handleTestRunStarted();
        }
        if (message.testCase) {
            this.handleTestCase(message.testCase);
        }
        if (message.testRunHookFinished) {
            this.handleTestRunHookFinished(message.testRunHookFinished);
        }
        if (message.testCaseStarted) {
            this.handleTestCaseStarted();
        }
        if (message.testStepFinished) {
            this.handleTestStepFinished();
        }
        if (message.testCaseFinished) {
            this.handleTestCaseFinished(message.testCaseFinished);
        }
        if (message.testRunFinished) {
            this.handleTestRunFinished(message.testRunFinished);
        }
    }
    handleUndefinedParameterType(undefinedParameterType) {
        this.pendingProblems.push({
            type: ProblemType.PARAMETER_TYPE,
            details: formatUndefinedParameterType(undefinedParameterType),
        });
    }
    handleTestRunStarted() {
        this.interceptor.acquire();
        this.rerender(true);
    }
    handleTestCase(testCase) {
        this.totalScenarios++;
        this.totalSteps += testCase.testSteps.length;
        this.rerender();
    }
    handleTestRunHookFinished(testRunHookFinished) {
        if (!NON_REPORTABLE_STATUSES.includes(testRunHookFinished.result.status)) {
            this.pendingProblems.push({
                type: ProblemType.GLOBAL_HOOK,
                details: composeGlobalHookSummary(testRunHookFinished, this.query, this.options, this.stream),
            });
            this.rerender();
        }
    }
    handleTestCaseStarted() {
        this.runningScenarios++;
        this.phase = Phase.RUNNING;
        this.rerender();
    }
    handleTestStepFinished() {
        this.finishedSteps++;
        this.rerender();
    }
    handleTestCaseFinished(testCaseFinished) {
        this.runningScenarios--;
        this.finishedScenarios++;
        if (testCaseFinished.willBeRetried) {
            const testCase = this.query.findTestCaseBy(testCaseFinished);
            if (testCase) {
                this.finishedScenarios--;
                this.finishedSteps = this.finishedSteps - testCase.testSteps.length;
            }
        }
        else {
            const mostSevereResult = this.query.findMostSevereTestStepResultBy(testCaseFinished);
            if (mostSevereResult && !NON_REPORTABLE_STATUSES.includes(mostSevereResult.status)) {
                this.pendingProblems.push({
                    type: ProblemType.TEST_CASE,
                    details: composeScenarioSummary(testCaseFinished, this.query, this.options, this.stream),
                });
            }
        }
        this.rerender();
    }
    handleTestRunFinished(testRunFinished) {
        this.phase = Phase.DONE;
        if (testRunFinished.exception) {
            const error = testRunFinished.exception?.stackTrace ||
                testRunFinished.exception?.message ||
                'Unknown error';
            this.pendingProblems.push({
                type: ProblemType.TEST_RUN,
                details: '\n' +
                    indent(formatError(error, TestStepResultStatus.FAILED, this.options.theme, this.stream), 4),
            });
        }
        this.rerender();
        this.interceptor.release();
    }
    rerender(initial = false) {
        let output = '';
        if (this.pendingProblems.length > 0) {
            if (this.printedProblems.length === 0) {
                output += 'Problems:\n';
            }
            const problems = this.pendingProblems.splice(0);
            for (const { type, details } of problems) {
                const prefixed = formatProblem(type, details, this.options.theme, this.stream);
                const number = this.printedProblems.length + 1;
                output += `${indentNumbered(prefixed, 2, number)}\n`;
                this.printedProblems.push({ type, details });
            }
        }
        if (this.phase === Phase.DONE && this.options.summarise) {
            output += this.makeSummaryBlock();
        }
        else {
            output += this.makeProgressBlock();
        }
        this.render(output, initial);
    }
    makeSummaryBlock() {
        let output = '\n';
        output += composeStats(this.query, this.options.theme, this.stream);
        output += '\n';
        const suggestions = findAllSuggestions(this.query);
        if (suggestions.length > 0) {
            output += composeSnippets(suggestions, this.options.formatCode, this.stream);
            output += '\n';
        }
        return output;
    }
    makeProgressBlock() {
        return [
            '',
            this.makeBar(this.finishedScenarios, this.totalScenarios, 'scenarios'),
            this.makeBar(this.finishedSteps, this.totalSteps, 'steps'),
            this.makeStatus(),
            '',
        ].join('\n');
    }
    makeBar(finished, total, label) {
        const barWidth = Math.min(this.stream.columns - MIN_LEGEND_WIDTH, MAX_BAR_WIDTH);
        const ratio = total > 0 ? finished / total : 0;
        const filledCount = Math.round(ratio * barWidth);
        const emptyCount = barWidth - filledCount;
        const filled = '█'.repeat(filledCount);
        const empty = '░'.repeat(emptyCount);
        return `${filled}${empty} ${finished}/${total} ${label}`;
    }
    makeStatus() {
        switch (this.phase) {
            case Phase.PREPARING:
                return 'Getting ready...';
            case Phase.RUNNING:
                return `Running ${this.runningScenarios} scenarios...`;
            case Phase.DONE:
                return 'Done';
        }
    }
    render(content, initial) {
        this.interceptor.bypass(() => {
            if (!initial) {
                this.stream.moveCursor(0, -4);
                this.stream.clearScreenDown();
            }
            this.stream.write(content);
        });
    }
}
