// @cucumber/pretty-formatter/dist/ProgressPrinter.js
import { Query } from '@cucumber/query';
import { defaultFormatCode } from './defaultFormatCode.js';
import { formatStatusCharacter } from './formatting/index.js';
import { SummaryPrinter } from './SummaryPrinter.js';
import { CUCUMBER_THEME } from './theme.js';
const DEFAULT_OPTIONS = {
    includeAttachments: true,
    summarise: false,
    theme: CUCUMBER_THEME,
    formatCode: defaultFormatCode,
};
/**
 * Prints test progress as single-character status indicators
 *
 * @remarks
 * Outputs a character for each test step as they complete, providing a compact
 * real-time view of test progress. Characters are styled according to their status
 * (e.g. `.` for passed, `F` for failed etc).
 */
export class ProgressPrinter {
    stream;
    print;
    options;
    query = new Query();
    /**
     * Creates a new ProgressPrinter instance
     *
     * @param params - Initialisation object
     * @param params.stream - The stream being written to, used for feature detection
     * @param params.options - Options for the output
     */
    constructor({ stream = process.stdout, options = {}, } = {}) {
        this.stream = stream;
        this.print = (content) => stream.write(content);
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options,
        };
    }
    /**
     * Processes a Cucumber message envelope and prints if appropriate
     *
     * @param message - The Cucumber message envelope to process
     */
    update(message) {
        this.query.update(message);
        if (message.testStepFinished) {
            this.print(formatStatusCharacter(message.testStepFinished.testStepResult.status, this.options.theme, this.stream));
        }
        if (message.testRunHookFinished) {
            this.print(formatStatusCharacter(message.testRunHookFinished.result.status, this.options.theme, this.stream));
        }
        if (message.testRunFinished) {
            this.print('\n');
            if (this.options.summarise) {
                this.summarise();
            }
        }
    }
    summarise() {
        SummaryPrinter.summarise(this.query, {
            stream: this.stream,
            options: this.options,
        });
    }
}
