// @cucumber/pretty-formatter/dist/SummaryPrinter.js
import { TestStepResultStatus } from '@cucumber/messages';
import { Query } from '@cucumber/query';
import { composeGlobalHookSummary, composeScenarioSummary, composeSnippets, composeStats, } from './composition/index.js';
import { defaultFormatCode } from './defaultFormatCode.js';
import { formatError, formatForStatus, formatStatusName, formatUndefinedParameterType, } from './formatting/index.js';
import { findAllSuggestions, findAllTestCaseFinishedInCanonicalOrder } from './queries/index.js';
import { CUCUMBER_THEME } from './theme.js';
import { GHERKIN_INDENT_LENGTH, indent, indentNumbered, NON_REPORTABLE_STATUSES, ORDERED_STATUSES, } from './utils.js';
const DEFAULT_OPTIONS = {
    includeAttachments: true,
    theme: CUCUMBER_THEME,
    formatCode: defaultFormatCode,
};
/**
 * Prints a summary of test results including non-passing scenarios, statistics, and snippets
 *
 * @remarks
 * Outputs non-passing scenarios and hooks, scenario and step counts, durations, and suggested
 * snippets for undefined steps.
 */
export class SummaryPrinter {
    stream;
    print;
    println;
    options;
    query = new Query();
    /**
     * Creates a new SummaryPrinter instance
     *
     * @param params - Initialisation object
     * @param params.stream - The stream being written to, used for feature detection
     * @param params.options - Options for the output
     */
    constructor({ stream = process.stdout, options = {}, } = {}) {
        this.stream = stream;
        this.print = (content) => stream.write(content);
        this.println = (content = '') => this.print(`${content}\n`);
        this.options = {
            ...DEFAULT_OPTIONS,
            ...options,
        };
    }
    /**
     * Processes a Cucumber message envelope and prints if appropriate
     *
     * @param envelope - The Cucumber message envelope to process
     */
    update(envelope) {
        this.query.update(envelope);
        if (envelope.testRunFinished) {
            this.printSummary();
        }
    }
    /**
     * Immediately print a summary from pre-populated Query object
     *
     * @param query - A pre-populated Query object to be summarised
     * @param params - See constructor
     */
    static summarise(query, { stream = process.stdout, options = {}, } = {}) {
        const printer = new SummaryPrinter({
            stream,
            options,
        });
        printer.query = query;
        printer.printSummary();
    }
    printSummary() {
        this.printNonPassingScenarios();
        this.printUnknownParameterTypes();
        this.printNonPassingGlobalHooks();
        this.printNonPassingTestRun();
        this.printStats();
        this.printSnippets();
    }
    printStats() {
        this.println();
        this.println(composeStats(this.query, this.options.theme, this.stream));
    }
    printNonPassingScenarios() {
        const nonPassingScenarios = this.resolveNonPassingScenarios();
        for (const status of ORDERED_STATUSES) {
            const forThisStatus = nonPassingScenarios.get(status) ?? [];
            if (forThisStatus.length > 0) {
                this.println();
                this.println(formatForStatus(status, `${formatStatusName(status)} scenarios:`, this.options.theme, this.stream));
                forThisStatus.forEach((testCaseFinished, index) => {
                    const formatted = composeScenarioSummary(testCaseFinished, this.query, this.options, this.stream);
                    this.println(indentNumbered(formatted, 2, index + 1));
                });
            }
        }
    }
    resolveNonPassingScenarios() {
        const reportableByStatus = new Map();
        for (const testCaseFinished of findAllTestCaseFinishedInCanonicalOrder(this.query)) {
            const mostSevereResult = this.query.findMostSevereTestStepResultBy(testCaseFinished);
            if (mostSevereResult && !NON_REPORTABLE_STATUSES.includes(mostSevereResult.status)) {
                if (!reportableByStatus.has(mostSevereResult.status)) {
                    reportableByStatus.set(mostSevereResult.status, []);
                }
                reportableByStatus.get(mostSevereResult.status)?.push(testCaseFinished);
            }
        }
        return reportableByStatus;
    }
    printUnknownParameterTypes() {
        const unknownParameterTypes = this.query.findAllUndefinedParameterTypes();
        if (unknownParameterTypes.length > 0) {
            this.println();
            this.println(formatForStatus(TestStepResultStatus.UNDEFINED, 'These parameters are missing a parameter type definition:', this.options.theme, this.stream));
            unknownParameterTypes.forEach((upt, index) => {
                this.println(indentNumbered(formatUndefinedParameterType(upt), GHERKIN_INDENT_LENGTH, index + 1));
            });
        }
    }
    printNonPassingGlobalHooks() {
        const testRunHookFinished = this.query.findAllTestRunHookFinished();
        const failedHooks = testRunHookFinished.filter((hook) => hook.result.status === TestStepResultStatus.FAILED);
        if (failedHooks.length > 0) {
            this.println();
            this.println(formatForStatus(TestStepResultStatus.FAILED, `${formatStatusName(TestStepResultStatus.FAILED)} hooks:`, this.options.theme, this.stream));
            failedHooks.forEach((testRunHookFinished, index) => {
                const formatted = composeGlobalHookSummary(testRunHookFinished, this.query, this.options, this.stream);
                this.println(indentNumbered(formatted, GHERKIN_INDENT_LENGTH, index + 1));
            });
        }
    }
    printNonPassingTestRun() {
        const testRunFinished = this.query.findTestRunFinished();
        if (testRunFinished?.exception) {
            this.println(formatForStatus(TestStepResultStatus.FAILED, `${formatStatusName(TestStepResultStatus.FAILED)} test run:`, this.options.theme, this.stream));
            const error = testRunFinished.exception?.stackTrace || testRunFinished.exception?.message;
            if (error) {
                this.print(indent(formatError(error, TestStepResultStatus.FAILED, this.options.theme, this.stream), 7));
            }
        }
    }
    printSnippets() {
        const suggestions = findAllSuggestions(this.query);
        if (suggestions.length > 0) {
            this.println(composeSnippets(suggestions, this.options.formatCode, this.stream));
        }
    }
}
