// @cucumber/pretty-formatter/dist/utils.js
import { stripVTControlCharacters } from 'node:util';
import { TestStepResultStatus } from '@cucumber/messages';
export const GHERKIN_INDENT_LENGTH = 2;
export const STEP_ARGUMENT_INDENT_LENGTH = 2;
export const ATTACHMENT_INDENT_LENGTH = 4;
export const ERROR_INDENT_LENGTH = 4;
export const ORDERED_STATUSES = [
    TestStepResultStatus.UNKNOWN,
    TestStepResultStatus.PASSED,
    TestStepResultStatus.SKIPPED,
    TestStepResultStatus.PENDING,
    TestStepResultStatus.UNDEFINED,
    TestStepResultStatus.AMBIGUOUS,
    TestStepResultStatus.FAILED,
];
export const NON_REPORTABLE_STATUSES = [TestStepResultStatus.PASSED, TestStepResultStatus.SKIPPED];
export var ProblemType;
(function (ProblemType) {
    ProblemType[ProblemType["PARAMETER_TYPE"] = 0] = "PARAMETER_TYPE";
    ProblemType[ProblemType["GLOBAL_HOOK"] = 1] = "GLOBAL_HOOK";
    ProblemType[ProblemType["TEST_CASE"] = 2] = "TEST_CASE";
    ProblemType[ProblemType["TEST_RUN"] = 3] = "TEST_RUN";
})(ProblemType || (ProblemType = {}));
export function ensure(value, message) {
    if (!value) {
        throw new Error(message);
    }
    return value;
}
export function join(...originals) {
    return originals.filter((part) => !!part).join(' ');
}
export function indent(original, by) {
    return original
        .split('\n')
        .map((line) => ' '.repeat(by) + line)
        .join('\n');
}
export function indentNumbered(original, by, number) {
    const baselineIndent = ' '.repeat(by);
    return original
        .split('\n')
        .map((line, i) => {
        if (line === '') {
            return '';
        }
        return i === 0 ? `${baselineIndent}${number}) ${line}` : `${baselineIndent}   ${line}`;
    })
        .join('\n');
}
export function pad(original) {
    return `\n${original}\n`;
}
export function unstyled(text) {
    return stripVTControlCharacters(text);
}
export function extractReportableMessage(testStepResult) {
    const { status, exception, message: standaloneMessage } = testStepResult;
    if (status === TestStepResultStatus.FAILED) {
        return exception?.stackTrace || exception?.message || standaloneMessage;
    }
    if (status === TestStepResultStatus.PENDING || status === TestStepResultStatus.SKIPPED) {
        return exception?.message || standaloneMessage;
    }
    return undefined;
}
