// @cucumber/html-formatter/dist/src/CucumberHtmlStream.js
import fs from 'node:fs';
import path from 'node:path';
import { Transform } from 'node:stream';
export class CucumberHtmlStream extends Transform {
    cssPath;
    jsPath;
    iconPath;
    template = null;
    preMessageWritten = false;
    postMessageWritten = false;
    firstMessageWritten = false;
    /**
     * @param cssPath
     * @param jsPath
     * @param iconPath
     */
    constructor(cssPath = path.join(import.meta.dirname, '..', 'main.css'), jsPath = path.join(import.meta.dirname, '..', 'main.js'), iconPath = path.join(import.meta.dirname, 'icon.url')) {
        super({ objectMode: true });
        this.cssPath = cssPath;
        this.jsPath = jsPath;
        this.iconPath = iconPath;
    }
    _transform(envelope, _encoding, callback) {
        if (this.postMessageWritten) {
            callback(new Error('Stream closed'));
            return;
        }
        this.writePreMessageUnlessAlreadyWritten((err) => {
            if (err) {
                callback(err);
                return;
            }
            this.writeMessage(envelope);
            callback();
        });
    }
    _flush(callback) {
        this.writePostMessage(callback);
    }
    writePreMessageUnlessAlreadyWritten(callback) {
        if (this.preMessageWritten) {
            return callback();
        }
        this.preMessageWritten = true;
        this.writeTemplateBetween(null, '{{title}}', (err) => {
            if (err) {
                return callback(err);
            }
            this.push('Cucumber');
            this.writeTemplateBetween('{{title}}', '{{icon}}', (err) => {
                if (err) {
                    return callback(err);
                }
                this.writeFile(this.iconPath, (err) => {
                    if (err) {
                        return callback(err);
                    }
                    this.writeTemplateBetween('{{icon}}', '{{css}}', (err) => {
                        if (err) {
                            return callback(err);
                        }
                        this.writeFile(this.cssPath, (err) => {
                            if (err) {
                                return callback(err);
                            }
                            this.writeTemplateBetween('{{css}}', '{{custom_css}}', (err) => {
                                if (err) {
                                    return callback(err);
                                }
                                this.writeTemplateBetween('{{custom_css}}', '{{messages}}', (err) => {
                                    if (err) {
                                        return callback(err);
                                    }
                                    callback();
                                });
                            });
                        });
                    });
                });
            });
        });
    }
    writePostMessage(callback) {
        this.writePreMessageUnlessAlreadyWritten((err) => {
            if (err) {
                return callback(err);
            }
            this.writeTemplateBetween('{{messages}}', '{{script}}', (err) => {
                if (err) {
                    return callback(err);
                }
                this.writeFile(this.jsPath, (err) => {
                    if (err) {
                        return callback(err);
                    }
                    this.writeTemplateBetween('{{script}}', '{{custom_script}}', (err) => {
                        if (err) {
                            return callback(err);
                        }
                        this.writeTemplateBetween('{{custom_script}}', null, callback);
                    });
                });
            });
        });
    }
    writeFile(path, callback) {
        const cssStream = fs.createReadStream(path, { encoding: 'utf-8' });
        cssStream.on('data', (chunk) => this.push(chunk));
        cssStream.on('error', (err) => callback(err));
        cssStream.on('end', callback);
    }
    writeTemplateBetween(begin, end, callback) {
        this.readTemplate((err, template) => {
            if (err) {
                return callback(err);
            }
            if (!template) {
                return callback(new Error('template is required if error is missing'));
            }
            const beginIndex = begin == null ? 0 : template.indexOf(begin) + begin.length;
            const endIndex = end == null ? template.length : template.indexOf(end);
            this.push(template.substring(beginIndex, endIndex));
            callback();
        });
    }
    readTemplate(callback) {
        if (this.template !== null) {
            return callback(null, this.template);
        }
        fs.readFile(`${import.meta.dirname}/index.mustache`, { encoding: 'utf-8' }, (err, template) => {
            if (err) {
                return callback(err);
            }
            this.template = template;
            return callback(null, template);
        });
    }
    writeMessage(envelope) {
        if (!this.firstMessageWritten) {
            this.firstMessageWritten = true;
        }
        else {
            this.push(',');
        }
        // Replace < with \x3C
        // https://html.spec.whatwg.org/multipage/scripting.html#restrictions-for-contents-of-script-elements
        this.push(JSON.stringify(envelope).replace(/</g, '\\x3C'));
    }
}
