// @cucumber/query/dist/Lineage.js
export var NamingStrategyLength;
(function (NamingStrategyLength) {
    NamingStrategyLength["LONG"] = "LONG";
    NamingStrategyLength["SHORT"] = "SHORT";
})(NamingStrategyLength || (NamingStrategyLength = {}));
export var NamingStrategyFeatureName;
(function (NamingStrategyFeatureName) {
    NamingStrategyFeatureName["INCLUDE"] = "INCLUDE";
    NamingStrategyFeatureName["EXCLUDE"] = "EXCLUDE";
})(NamingStrategyFeatureName || (NamingStrategyFeatureName = {}));
export var NamingStrategyExampleName;
(function (NamingStrategyExampleName) {
    NamingStrategyExampleName["NUMBER"] = "NUMBER";
    NamingStrategyExampleName["PICKLE"] = "PICKLE";
    NamingStrategyExampleName["NUMBER_AND_PICKLE_IF_PARAMETERIZED"] = "NUMBER_AND_PICKLE_IF_PARAMETERIZED";
})(NamingStrategyExampleName || (NamingStrategyExampleName = {}));
class BuiltinNamingStrategy {
    length;
    featureName;
    exampleName;
    constructor(length, featureName, exampleName) {
        this.length = length;
        this.featureName = featureName;
        this.exampleName = exampleName;
    }
    reduce(lineage, pickle) {
        const parts = [];
        if (lineage.feature?.name && this.featureName === NamingStrategyFeatureName.INCLUDE) {
            parts.push(lineage.feature.name);
        }
        if (lineage.rule?.name) {
            parts.push(lineage.rule.name);
        }
        if (lineage.scenario?.name) {
            parts.push(lineage.scenario.name);
        }
        else {
            parts.push(pickle.name);
        }
        if (lineage.examples?.name) {
            parts.push(lineage.examples.name);
        }
        if (lineage.example) {
            const exampleNumber = [
                '#',
                (lineage.examplesIndex ?? 0) + 1,
                '.',
                (lineage.exampleIndex ?? 0) + 1,
            ].join('');
            switch (this.exampleName) {
                case NamingStrategyExampleName.NUMBER:
                    parts.push(exampleNumber);
                    break;
                case NamingStrategyExampleName.NUMBER_AND_PICKLE_IF_PARAMETERIZED:
                    if (lineage.scenario?.name !== pickle.name) {
                        parts.push(`${exampleNumber}: ${pickle.name}`);
                    }
                    else {
                        parts.push(exampleNumber);
                    }
                    break;
                case NamingStrategyExampleName.PICKLE:
                    parts.push(pickle.name);
                    break;
            }
        }
        if (this.length === NamingStrategyLength.SHORT) {
            return parts.at(-1);
        }
        return parts.filter((part) => !!part).join(' - ');
    }
}
export function namingStrategy(length, featureName = NamingStrategyFeatureName.INCLUDE, exampleName = NamingStrategyExampleName.NUMBER_AND_PICKLE_IF_PARAMETERIZED) {
    return new BuiltinNamingStrategy(length, featureName, exampleName);
}
