// @cucumber/message-streams/dist/AttachmentExternalisingStream.js
import { writeFile } from 'node:fs/promises';
import path from 'node:path';
import { Transform } from 'node:stream';
import { AttachmentContentEncoding, IdGenerator, } from '@cucumber/messages';
import mime from 'mime';
const alwaysInlinedTypes = ['text/x.cucumber.log+plain', 'text/uri-list'];
const encodingsMap = {
    IDENTITY: 'utf-8',
    BASE64: 'base64',
};
/**
 * A transform stream that externalises attachment bodies from a stream of
 * {@link Envelope}s by writing them to files in a given directory and
 * replacing each body with a relative URL pointing to the written file.
 */
export class AttachmentExternalisingStream extends Transform {
    writeOperations = [];
    options;
    constructor(options) {
        super({ objectMode: true });
        this.options = {
            behaviour: false,
            ...options,
            newId: options.newId ?? IdGenerator.uuid(),
        };
    }
    _transform(envelope, _encoding, callback) {
        if (envelope.attachment && this.shouldExternalise(envelope.attachment)) {
            const { attachment, writeOperation } = rewriteAttachment(envelope.attachment, this.options.directory, this.options.newId);
            this.push({ ...envelope, attachment });
            if (writeOperation) {
                this.writeOperations.push(writeOperation);
            }
        }
        else {
            this.push(envelope);
        }
        callback();
    }
    shouldExternalise(attachment) {
        const { behaviour } = this.options;
        if (Array.isArray(behaviour)) {
            return (!alwaysInlinedTypes.includes(attachment.mediaType) &&
                behaviour.some((pattern) => matchMediaType(attachment.mediaType, pattern)));
        }
        return behaviour === true;
    }
    _flush(callback) {
        Promise.all(this.writeOperations).then(() => callback(), (err) => callback(err));
    }
}
function matchMediaType(mediaType, pattern) {
    const [patternType, patternSubtype] = pattern.split('/');
    const [type, subtype] = mediaType.split('/');
    return patternType === type && (patternSubtype === '*' || patternSubtype === subtype);
}
function rewriteAttachment(original, directory, newId) {
    if (alwaysInlinedTypes.includes(original.mediaType)) {
        return { attachment: original };
    }
    let filename = `attachment-${newId()}`;
    const extension = mime.getExtension(original.mediaType);
    if (extension) {
        filename += `.${extension}`;
    }
    const writeOperation = writeFile(path.join(directory, filename), Buffer.from(original.body, encodingsMap[original.contentEncoding]));
    const attachment = {
        ...original,
        contentEncoding: AttachmentContentEncoding.IDENTITY,
        body: '',
        url: `./${filename}`,
    };
    return { attachment, writeOperation };
}
